docs: draft sub-path routing docs (#7422)
vuestorefront/vue-storefront: docs/content/guides/7.multistore/3.patterns/5.subpath/1.subpath.md

Reviewer: I feel like ### How Config Switcher Works and ## Setting Up the Middleware section could be described in our [guide on Config Switcher](https://github.com/vuestorefront/enterprise/blob/b7a086a160043c4f56168c0526db35f0201e5f3a/docs/enterprise/connect/middleware/2.guides/6.config-switcher.md)

---
title: How to Make Sub-path Routing
layout: default
navigation:
  icon: tabler:number-1-small
---

# How to Make Sub-path Routing

> Sub-path routing allows you to serve different store configurations from the same Alokai deployment using URL paths. This approach is ideal when you need to maintain multiple stores with different configurations but want to manage them from a single codebase.

Reviewer 2: ```suggestion
Sub-path routing allows you to serve different store configurations from the same Alokai deployment using URL paths. This approach is useful if the multibrand per-domain does not address your requirements.
```

    - Different configuration ID set by different paths allow to fetch data from different sources
    - Example: `/apparel` path shows apparel products, while `/electronics` shows electronics products

2. **Configuration ID vs Store ID**
    - Each configuration has its own distinct ID
    - This is different from the store ID (`storeId`)
    - Multiple configurations can exist within a single store

3. **Same-Store Operation**
    - Config Switcher works within a single store instance
    - It extracts the configuration ID from the request
    - It executes requests using the appropriate configuration

::tip Config Switcher Documentation
This section provides a brief overview of setting up Config Switcher. For complete documentation, refer to the [Config Switcher guide](https://docs.alokai.com/middleware/guides/config-switcher).
::

### Key Distinctions To File-Based Inheritance

`ConfigSwitcher` operates differently from file-based inheritance:

| **File-Based Inheritance**              | **Config Switcher**                                              |
| --------------------------------------- | ---------------------------------------------------------------- |

> | Works **between different stores**      | Works **within a single store**                                  |

Reviewer 2: ```suggestion
| Creates instance per store      | Switches between various configurations within a single instance                                  |
```

> | Works **between different stores**      | Works **within a single store**                                  |

Author: done

| Creates separate deployable instances   | Uses a single deployable instance                                |

> | Full codebase customization             | Only middleware configuration changes, Storefront code is shared |

Reviewer 2: ```suggestion
| Provides inheritance mechanism to make managing multiple stores easy             | You have to manage codebase reuse between stores on your own |
```
It's quite confusing, especially that we then proceed with customising UI per config switcher's store.

import { createConfigSwitcherExtension } from '@alokai/connect/config-switcher';
import type { MiddlewareConfig } from '@vsf-enterprise/<integration>-api';

export const configSwitcherExtension = createConfigSwitcherExtension<MiddlewareConfig>({
  // Config switcher options - specify only what's different for each store
  configuration: {
    apparel: {
      api: {
        baseSiteId: 'apparel',
        catalogId: 'apparelProductCatalog',
      },
    },
    electronics: {
      api: {
        baseSiteId: 'electronics',
        catalogId: 'electronicsProductCatalog',
      },
    },
  },
});
```

Then add this extension to your integration configuration:

```ts
// apps/storefront-middleware/integrations/<integration-name>/config.ts
import { configSwitcherExtension } from './extensions/configSwitcher';

export default {
  integrations: {
    <integration-name>: {
      // Base configuration shared across all stores
      configuration: {
        // ...
      },
      extensions: (predefinedExtensions) => [
        ...predefinedExtensions,
        configSwitcherExtension,
      ]
    }
  }
};
```

> ## Customizing Store Appearance with CSS Variables

Reviewer 2: Please add a complete example on how to do it

Author: done https://github.com/vuestorefront/vue-storefront/pull/7422/commits/c8e0ca6d5af95e822a75a0d9dcc736f604b215f3

export default defineSdkConfig(({ buildModule, config, getRequestHeaders, middlewareModule }) => {
  const route = useRoute();
  
  // Create a reusable function for determining the config ID
  function getConfigSwitcherHeader() {
    return route.path.includes('/electronics') ? 'electronics' : 'apparel';
  }
  
  return {
    // Unified commerce module
    unified: buildModule(middlewareModule<UnifiedEndpoints>, {
      apiUrl: `${config.apiUrl}/commerce/unified`,
      defaultRequestConfig: {
        getConfigSwitcherHeader,
        headers: getRequestHeaders(),
      },
      // ... other config
    }),
    // ... other modules
  };
});
```
::

## Updating the File Structure for Path Routing

To enable path-based routing, you need to restructure your application to include a dynamic `[configId]` parameter in the route. This allows you to access the store identifier directly from the URL.

::tabs{:titles='["Next.js", "Nuxt"]'}

#tab-1
For Next.js apps using the App Router, update your file structure:

```bash
apps/storefront-unified-nextjs/app/
├── [locale]/                 # Language parameter

> │   ├── [configId]/           # Store identifier parameter

Reviewer 2: Do we have anything against having the order reversed? If no, then please add an information that this can go both ways

Author: it didn't work in Next.js - at least without any other changes in the i18n config

Reviewer 2: Ty, I've created a spike for the next sprint, unfortunatelly there's a business case for this

│   ├── products/            # Product pages
│   │   ├── index.vue
│   │   └── [id].vue
│   ├── some-electronics-page.vue # Electronics-specific page
│   └── some-apparel-page.vue     # Apparel-specific page
```
::

This structure enables routes like:
- `/en/electronics` - Electronics store home
- `/en/apparel` - Apparel store home
- `/en/electronics/category` - PLP in the electronics store
- `/en/apparel/category` - PLP in the apparel store

## Updating Internal Links

::warning Update Internal Links
After setting up the path routing structure, you must update all internal links in your application to include the appropriate configId parameter. This is crucial to maintain proper navigation within each store and prevent users from accidentally switching between stores when clicking links.
::

Make sure that all your internal links and navigation components preserve the store context by including the current configId in the URLs. This applies to:

- Navigation menus
- Product links
- Category links
- Cart and checkout flows
- Footer links
- Breadcrumbs
- Any other internal navigation elements

Without this, users might unexpectedly switch between stores while browsing your site.

> ## Applying Store-Specific Styling

Reviewer 2: Group this with CSS section as it is too conceptually scattered for the reader.

Author: done https://github.com/vuestorefront/vue-storefront/pull/7422/commits/c8e0ca6d5af95e822a75a0d9dcc736f604b215f3

Now that you have CSS variables defined for each store, you need to apply the store identifier as a CSS class to enable the different styles.

::tabs{:titles='["Next.js", "Nuxt"]'}

#tab-1
In your Next.js layout, apply the configId as a class:

```tsx
// apps/storefront-unified-nextjs/app/[locale]/[configId]/layout.tsx
import { PropsWithChildren } from 'react';

interface LayoutProps extends PropsWithChildren {
  params: {
    configId: string;
  };
}

export default function Layout({ children, params: { configId } }: LayoutProps) {
  return (
    <div className={configId}>
      {children}
    </div>
  );
}
```

#tab-2
In your Nuxt app.vue, apply the configId as a class:

```vue
// apps/storefront-unified-nuxt/app.vue
<template>

>   <Body :class="['font-body', route.params.configId]" />

Reviewer 2: beautiful 💖

<script setup lang="ts">
const route = useRoute();
</script>
```
::

## Styling Components for Different Stores

With the store identifier applied as a class, you can now customize components with store-specific styles. You'll need to update your Tailwind configuration to support this.

::tabs{:titles='["Next.js", "Nuxt"]'}

#tab-1
First, update your Tailwind config:

```ts
// packages-end-user/tailwind-config/src/nextjs.ts
import { tailwindConfig } from "@storefront-ui/react/tailwind-config";
import type { Config } from "tailwindcss";
import plugin from "tailwindcss/plugin";

const config: Config = {
  // ...
  plugins: [
    // ...
    plugin(({ addVariant }) => {
      // Add variant for electronics store
      addVariant("electronics", ".electronics &");
      // Add variant for apparel store
      addVariant("apparel", ".apparel &");
    }),
  ],
  // ...
};
export default config;
```

Then use these variants (`electronics:`, `apparel:` etc.) in your components:

```tsx
// apps/storefront-unified-nextjs/components/navigations/navbar-top.tsx
export default function NavbarTop({ children, className, filled }: NavbarTopProps) {
  return (
    <header
      className={classNames(

>         'sticky top-0 z-40 flex h-14 bg-primary-800 electronics:!bg-primary-600 apparel:!bg-primary-400 md:-top-5 md:h-20 md:pt-2.5',

Reviewer 2: I didn't know of this tailwind feature, will this example work? You have 3 backgrounds defined here? Do I understand correctly that it will try 800 then overwrite this with one of the variants if they are available? But then Idk how does it know which variant is the active one. Could you make it a bit more clear in the guide?

Author: it works, I've added short explanation

![CleanShot 2025-04-07 at 17 02 23@2x](https://github.com/user-attachments/assets/f9a20d08-71c5-4ed9-9533-7abbee98fdfe)

        // other classes...
      )}
      data-testid="navbar-top"
    >
      {/* Component content */}
    </header>
  );
}
```

#tab-2
First, update your Tailwind config:

```ts
// packages-end-user/tailwind-config/src/nuxt.ts
import type { Config } from "tailwindcss";
import plugin from "tailwindcss/plugin";

export default {
  // ...
  plugins: [
    // ...
    plugin(({ addVariant }) => {
      // Add variant for electronics store
      addVariant("electronics", ".electronics &");
      // Add variant for apparel store
      addVariant("apparel", ".apparel &");
    }),
  ],
  // ...
} as Config;
```

Then use these variants (`electronics:`, `apparel:` etc.) in your components:

```vue
// apps/storefront-unified-nuxt/components/ui/NavbarTop/NavbarTop.vue
<template>
  <header
    :class="[
      'h-14 md:h-20 flex z-40 sticky top-0 md:-top-5 md:pt-2.5 md:shadow-md',

>       { 'bg-primary-700 electronics:bg-primary-500 apparel:bg-primary-300 text-white': filled },

Reviewer 2: Same as [here](https://github.com/vuestorefront/vue-storefront/pull/7422/files#r2030882285)

::

### Key Distinctions To File-Based Inheritance

`ConfigSwitcher` operates differently from file-based inheritance:

| **File-Based Inheritance**              | **Config Switcher**                                              |
| --------------------------------------- | ---------------------------------------------------------------- |
| Creates separate deployable instances   | Uses a single deployable instance                                |
| Provides inheritance mechanism to make managing multiple stores easy             | You have to manage codebase reuse between stores on your own |
| Requires separate builds for each store | Single build supports multiple configurations                    |

Take a look at the diagram to see the difference:

<img src="../../img/file-based-vs-switcher-diagram.png" alt="Comparison of File-based Inheritance vs Config Switcher" class="mx-auto">

## Setting Up the Middleware

The first step is to configure your Middleware to support different configurations using the Config Switcher extension.

> Create a Config Switcher extension for your integration:

Reviewer: our ::steps:: component would fit perfectly here

> ```ts

Reviewer: I think this would be the correct syntax for adding file path to a fenced code block
```suggestion
```ts{}[apps/storefront-middleware/integrations/<integration-name>/extensions/configSwitcher.ts]
```

Author: I was using this syntax in the past, but other team members in multibrand team said that it's better to set the file path as a comment, because the UI element for the file path looks bad when the path is long. We didn't have capacity that time to improve this UI element, so I've decided to use the comment instead

// apps/storefront-middleware/integrations/<integration-name>/extensions/configSwitcher.ts
import { createConfigSwitcherExtension } from '@alokai/connect/config-switcher';
import type { MiddlewareConfig } from '@vsf-enterprise/<integration>-api';

export const configSwitcherExtension = createConfigSwitcherExtension<MiddlewareConfig>({
  // Config switcher options - specify only what's different for each store
  configuration: {
    apparel: {
      api: {
        baseSiteId: 'apparel',
        catalogId: 'apparelProductCatalog',
      },
    },
    electronics: {
      api: {
        baseSiteId: 'electronics',
        catalogId: 'electronicsProductCatalog',
      },
    },
  },
});
```

Then add this extension to your integration configuration:

```ts
// apps/storefront-middleware/integrations/<integration-name>/config.ts
import { configSwitcherExtension } from './extensions/configSwitcher';

> export default {

Reviewer: You are showing an example of an integration-specific config file but the example looks more like middleware.config.ts - I think that is what you meant:

```ts
export const config = {
  configuration: {},
  extensions: (extensions: ApiClientExtension[]) => [
    ...extensions,
  ],
  location: '@vsf-enterprise/<integration-name>/server',
} satisfies Integration<Config>;
```

  integrations: {
    <integration-name>: {
      // Base configuration shared across all stores
      configuration: {
        // ...
      },
      extensions: (predefinedExtensions) => [
        ...predefinedExtensions,
        configSwitcherExtension,
      ]
    }
  }
};
```

## Customizing Store Appearance with CSS Variables

To create distinct visual identities for each store, you can define CSS variables that change based on the store config ID. This is a powerful way to maintain a consistent component library while applying different themes.

::tabs{:titles='["Next.js", "Nuxt"]'}

#tab-1
In your Next.js app, add custom CSS variables in your global styles:

> ```scss

Reviewer: do you think it would make sense to turn this code example into a diff?

// apps/storefront-unified-nextjs/app/[locale]/globals.scss
@tailwind base;
@tailwind components;
@tailwind utilities;

// Base styles...

// Electronics store theme
.electronics {
  --colors-primary-50: 45 249 255;
  --colors-primary-100: 233 243 255;
  --colors-primary-200: 200 224 255;
  --colors-primary-300: 166 204 255;
  --colors-primary-400: 110 161 255;
  --colors-primary-500: 51 117 255;
  --colors-primary-600: 46 106 230;
  --colors-primary-700: 38 78 191;
  --colors-primary-800: 29 63 153;
  --colors-primary-900: 176 196 244;
}

// Apparel store theme (different color scheme)
.apparel {
  --colors-primary-50: 243 254 249;
  --colors-primary-100: 224 247 235;
  --colors-primary-200: 187 235 210;
  --colors-primary-300: 135 216 177;
  --colors-primary-400: 77 192 140;
  --colors-primary-500: 45 165 116;
  --colors-primary-600: 34 134 95;
  --colors-primary-700: 31 110 80;
  --colors-primary-800: 30 86 65;
  --colors-primary-900: 23 64 49;
}
```

#tab-2

> In your Nuxt app, add custom CSS variables in your style.scss:

Reviewer: ```suggestion
In your Nuxt app, add custom CSS variables in your `style.scss`:
```

::

::tip Understanding Generated CSS
When using store-specific variants like `electronics:!bg-primary-600`, Tailwind generates scoped CSS rules. For example:

```css
.electronics .electronics\:\!bg-primary-600 {
    --tw-bg-opacity: 1 !important;
    background-color: rgb(46 106 230 / 1 !important) !important;
}
```

This ensures the styles only apply when two conditions are met: the parent element must have the `.electronics` class and the target element must have the `electronics:!bg-primary-600` class.
::

## Setting Up the Storefront

Now you need to configure your Storefront to send the appropriate config ID to the Middleware based on the current URL path.

::tabs{:titles='["Next.js", "Nuxt"]'}

#tab-1
In your Next.js SDK configuration, implement the `getConfigSwitcherHeader` function:

```ts
// apps/storefront-unified-nextjs/sdk/config.ts
import { defineSdkConfig, getPathnameFromRequestHeaders } from '@vue-storefront/next';
import type { UnifiedEndpoints } from 'storefront-middleware/types';

> function getConfigSwitcherHeader(baseHeaders: Record<string, string>) {

Reviewer: If we merge the PR with config switcher adjustments, remember to update the code example so that it uses the new `defineGetConfigSwitcherHeader` util

In your Nuxt SDK configuration, define a reusable `getConfigSwitcherHeader` function:

```ts
// apps/storefront-unified-nuxt/sdk.config.ts
import type { CommerceEndpoints, UnifiedCmsEndpoints, UnifiedEndpoints } from 'storefront-middleware/types';

export default defineSdkConfig(({ buildModule, config, getRequestHeaders, middlewareModule }) => {
  const route = useRoute();
  
  // Create a reusable function for determining the config ID
  function getConfigSwitcherHeader() {
    return route.path.includes('/electronics') ? 'electronics' : 'apparel';
  }
  
  return {
    // Unified commerce module
    unified: buildModule(middlewareModule<UnifiedEndpoints>, {
      apiUrl: `${config.apiUrl}/commerce/unified`,
      defaultRequestConfig: {
        getConfigSwitcherHeader,
        headers: getRequestHeaders(),
      },
      // ... other config
    }),
    // ... other modules
  };
});
```
::

## Updating the File Structure for Path Routing

To enable path-based routing, you need to restructure your application to include a dynamic `[configId]` parameter in the route. This allows you to access the store identifier directly from the URL.

::tabs{:titles='["Next.js", "Nuxt"]'}

#tab-1

> For Next.js apps using the App Router, update your file structure:

Reviewer: nice to have: seems like these file structure adjustments could be done with a single `mv` command run and we could add it to the guide here

│   │   └── layout.tsx        # Apply store-specific class
```

#tab-2
For Nuxt apps, update your file structure:

```bash
apps/storefront-unified-nuxt/pages/
├── [configId]/              # Store identifier parameter
│   ├── index.vue            # Home page
│   ├── cart/                # Cart pages
│   │   └── index.vue
│   ├── checkout/            # Checkout pages
│   │   └── index.vue
│   ├── products/            # Product pages
│   │   ├── index.vue
│   │   └── [id].vue
│   ├── some-electronics-page.vue # Electronics-specific page
│   └── some-apparel-page.vue     # Apparel-specific page
```
::

This structure enables routes like:
- `/en/electronics` - Electronics store home
- `/en/apparel` - Apparel store home
- `/en/electronics/category` - PLP in the electronics store
- `/en/apparel/category` - PLP in the apparel store

## Updating Internal Links

::warning Update Internal Links
After setting up the path routing structure, you must update all internal links in your application to include the appropriate configId parameter. This is crucial to maintain proper navigation within each store and prevent users from accidentally switching between stores when clicking links.
::

> Make sure that all your internal links and navigation components preserve the store context by including the current configId in the URLs. This applies to:

Reviewer: I would add one example showing how to update Next `<Link />` and Nuxt `<NuxtLink />` here to make it 100% clear what we mean

- Navigation menus
- Product links
- Category links
- Cart and checkout flows
- Footer links
- Breadcrumbs
- Any other internal navigation elements

Without this, users might unexpectedly switch between stores while browsing your site.

## Store-Specific Styling

There are two key steps to implement store-specific styling:

### 1. Applying Store Identifier Class

First, you need to apply the store identifier as a CSS class to enable different styles for each store:

::tabs{:titles='["Next.js", "Nuxt"]'}

#tab-1

> In your Next.js layout, apply the configId as a class:

Reviewer: the path to layout file sort-of suggests that but I think it is worth mentioning the underlying requirement here: the class has to be applied to an HTML element which is as high in the document hierarchy as possible so that we can style more deeply nested elements

Author: good idea, done :)

```tsx
// apps/storefront-unified-nextjs/app/[locale]/[configId]/layout.tsx
import { PropsWithChildren } from 'react';

interface LayoutProps extends PropsWithChildren {
  params: {
    configId: string;
  };
}

export default function Layout({ children, params: { configId } }: LayoutProps) {
  return (
    <div className={configId}>
      {children}
    </div>
  );
}
```

#tab-2
In your Nuxt app.vue, apply the configId as a class:

```vue
// apps/storefront-unified-nuxt/app.vue
<template>
  <Body :class="['font-body', route.params.configId]" />
  <NuxtLayout>
    <NuxtPage />
  </NuxtLayout>
</template>

<script setup lang="ts">
const route = useRoute();
</script>
```
::

### 2. Defining Store-Specific CSS Variables

Once the store identifier class is applied, you can define CSS variables that change based on the store:

::tabs{:titles='["Next.js", "Nuxt"]'}

#tab-1

> In your Next.js app, add custom CSS variables in your global styles:

Reviewer: wasn't it already described in lines 119-200?

Author: ah, sorry I've resolved the conflicts incorrectly

```ts
// packages-end-user/tailwind-config/src/nuxt.ts
import type { Config } from "tailwindcss";
import plugin from "tailwindcss/plugin";

export default {
  // ...
  plugins: [
    // ...
    plugin(({ addVariant }) => {
      // Add variant for electronics store
      addVariant("electronics", ".electronics &");
      // Add variant for apparel store
      addVariant("apparel", ".apparel &");
    }),
  ],
  // ...
} as Config;
```

Then use these variants (`electronics:`, `apparel:`, etc.) in your components:

```vue
// apps/storefront-unified-nuxt/components/ui/NavbarTop.vue
<template>
  <header
    :class="[
      'sticky top-0 z-40 flex h-14',
      'bg-primary-800',
      'electronics:!bg-primary-600',
      'apparel:!bg-primary-400',
      'md:-top-5 md:h-20 md:pt-2.5',
    ]"
  >
    <slot />
  </header>
</template>
```
::

::tip Using CSS Variables with Tailwind
The CSS variables we defined can be used with Tailwind's color system. For example, `bg-primary-500` will use the store-specific primary color defined in the CSS variables. For more information, see the [Tailwind CSS documentation](https://v3.tailwindcss.com/docs/customizing-colors#using-css-variables).
::

> ::tip Understanding Generated CSS

Reviewer: This one is also repeated (I already read it on line 206). However, this is the location where I would expect this tip so the first one should go and this one should stay 👍

Author: ah, sorry I've resolved the conflicts incorrectly

import { ApparelAnnouncementBar } from '@/components/store-apparel/announcement-bar';
import { ElectronicsAnnouncementBar } from '@/components/store-electronics/announcement-bar';

interface AnnouncementBarProps {
  configId: string;
}

function AnnouncementBar({ configId }: AnnouncementBarProps) {
  const StoreComponent = {
    apparel: ApparelAnnouncementBar,
    electronics: ElectronicsAnnouncementBar,
  }[configId];

  return StoreComponent ? <StoreComponent /> : null;
}

// Usage in a page or layout:
export default function Layout({ children, params: { configId } }: LayoutProps) {
  return (
    <div className={configId}>
      <AnnouncementBar configId={configId} />
      {children}
    </div>
  );
}
```

#tab-2
```vue
<!-- apps/storefront-unified-nuxt/components/AnnouncementBar.vue -->
<template>
  <component :is="StoreComponent" v-if="StoreComponent" />
</template>

<script setup lang="ts">
const props = defineProps<{
  configId: string;
}>();

const AnnouncementBarElectronics = resolveComponent('StoreElectronicsAnnouncementBar');
const AnnouncementBarApparel = resolveComponent('StoreApparelAnnouncementBar');

> const AnnouncementBar = computed(

Reviewer: suggestion: create a wrapper component to render dynamic component rendering based on `configId`. 

```ts
<StoreComponentRenderer
  components= {
    electronics: ElectronicsAnnouncementBar,
    apparel: ApparelAnnouncementBar
  }
/>
```

It could also be used in this scenario:
```
      {configId === 'electronics' && (
        <ElectronicsPromo />
      )}
```

like this:

```ts
<StoreComponentRenderer
  components= {
    electronics: ElectronicsPromo,
  }
/>
```
